src/System.Data.Common/src/System/Data/Common/DbDataReaderExtensions.cs

```diff
@@ -2,22 +2,97 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
+
 namespace System.Data.Common
 {
     public static class DbDataReaderExtensions
     {
         public static System.Collections.ObjectModel.ReadOnlyCollection<DbColumn> GetColumnSchema(this DbDataReader reader)
         {
-            if (reader.CanGetColumnSchema())
+            if (reader is null)
+            {
+                throw new ArgumentNullException(nameof(reader));
+            }
+            if (reader is IDbColumnSchemaGenerator schemaGenerator)
             {
-                return ((IDbColumnSchemaGenerator)reader).GetColumnSchema();
+                return schemaGenerator.GetColumnSchema();
             }
-            throw new NotSupportedException();
+            return GetColumnSchemaCompatibility(reader);
         }
 
         public static bool CanGetColumnSchema(this DbDataReader reader)
         {
-            return reader is IDbColumnSchemaGenerator;
+            return true;
+        }
+
+        private static System.Collections.ObjectModel.ReadOnlyCollection<DbColumn> GetColumnSchemaCompatibility(DbDataReader reader)
+        {
+            IList<DbColumn> columnSchema = new List<DbColumn>();
+            DataTable schemaTable = reader.GetSchemaTable();
+            DataColumnCollection schemaTableColumns = schemaTable.Columns;
+            foreach (DataRow row in schemaTable.Rows)
+            {
+                DbColumn dbColumn = new DataRowDbColumn(row, schemaTableColumns);
+                columnSchema.Add(dbColumn);
+            }
+            System.Collections.ObjectModel.ReadOnlyCollection<DbColumn> readOnlyColumnSchema = new System.Collections.ObjectModel.ReadOnlyCollection<DbColumn>(columnSchema);
+            return readOnlyColumnSchema;
+        }
+    }
+
+    internal sealed class DataRowDbColumn : DbColumn
```

Comment:
Why not move this into DbDataReaderExtensions, no reason for this to be internally visible...

Reply:
Done.